src/sync/mod.rs

```diff
@@ -0,0 +1,355 @@
+        {
+            keypair
+        } else {
+            return Ok(());
+        };
+        // Create `SignInput` transactions.
+        let redeem_script = config.redeem_script();
+        let block_height = match proposal.anchoring_payload() {
+            Some(payload) => payload.block_height,
+            None => {
+                return Err(AnchoringChainUpdateError::Internal(failure::format_err!(
+                    "Incorrect anchoring proposal found: {:?}",
+                    proposal
+                )))
+            }
+        };
+
+        log::info!(
+            "Found a new unfinished anchoring transaction proposal for height: {}",
+            block_height
+        );
+
+        let mut signer = p2wsh::InputSigner::new(redeem_script);
+        let sign_input_messages = inputs
+            .iter()
+            .enumerate()
+            .map(|(index, proposal_input)| {
+                let signature = signer.sign_input(
+                    TxInRef::new(proposal.as_ref(), index),
+                    inputs[index].as_ref(),
+                    &(keypair.1).0.key,
+                )?;
+
+                signer.verify_input(
+                    TxInRef::new(proposal.as_ref(), index),
+                    proposal_input.as_ref(),
+                    &(keypair.0).0,
+                    &signature,
+                )?;
+
+                Ok(SignInput {
+                    transaction: proposal.clone(),
+                    input: index as u32,
+                    input_signature: signature.into(),
+                })
+            })
+            .collect::<Result<Vec<_>, failure::Error>>()
+            .map_err(AnchoringChainUpdateError::Internal)?;
+        // Send sign input transactions to the Exonum node.
+        for sign_input in sign_input_messages {
+            self.api_client
+                .sign_input(sign_input)
+                .wait()
+                .map_err(AnchoringChainUpdateError::Client)?;
+        }
+        Ok(())
+    }
+
+    fn find_private_key(
+        &self,
+        anchoring_keys: impl IntoIterator<Item = btc::PublicKey>,
+    ) -> Option<(btc::PublicKey, btc::PrivateKey)> {
+        anchoring_keys.into_iter().find_map(|public_key| {
+            self.key_pool
+                .get(&public_key)
+                .cloned()
+                .map(|private_key| (public_key, private_key))
+        })
+    }
+}
+
+/// Errors that occur when updating the sync with Bitcoin task.
+#[derive(Debug)]
+pub enum SyncWithBitcoinError<C: Display, R: Display> {
+    /// Error occurred in the private API client.
+    Client(C),
+    /// Error occurred in the Bitcoin relay.
+    Relay(R),
+    /// Internal error.
+    Internal(failure::Error),
+    /// Initial funding transaction is unconfirmed.
+    UnconfirmedFundingTransaction(Hash),
+}
+
+/// Pushes anchoring transactions to the Bitcoin blockchain.
+#[derive(Debug)]
+pub struct SyncWithBitcoinTask<T, R>
```

Comment:
Is it intentional that the task does not maintain state (i.e., the last processed transaction index)? Could lead to better encapsulation, IMO.

Reply:
Stateless approach imo easier for testing.